src/Orleans/Messaging/Message.cs

```diff
@@ -601,7 +606,7 @@ private List<ArraySegment<byte>> Serialize_Impl(out int headerLengthOut, out int
             {
                 BufferPool.GlobalPool.Release(headerBytes);
             }
-            headerBytes = headerStream.ToBytes() as List<ArraySegment<byte>>;
+            headerBytes = context.Stream.ToBytes() as List<ArraySegment<byte>>;
             int headerLength = headerBytes.Sum(ab => ab.Count);
```

Comment:
It's a bit complicated to do this for the body, since it's not always set directly from a stream. We could avoid allocations by writing a helper function.

I consider this out of scope for this PR, though.